Automattic/wp-calypso
# composite-checkout: Split up useShoppingCart so cart management is separate

#### Changes proposed in this Pull Request

Composite checkout in calypso uses a hook called `useShoppingCart` to interact with the shopping-cart endpoint. However, it serves a dual purpose: it also translates the shopping cart data as returned by the endpoint (of type `ResponseCart`) into the data format required by the `@automattic/composite-checkout` package (type `WPCOMCart`).

This dual purpose has made the hook rather large and hard to reason about. It also makes it fairly bloated for applications outside of composite checkout, which we might want to explore doing in future PRs. 

In this PR we split that hook into two pieces: `useShoppingCartManager`, which just deals with the shopping-cart endpoint, and the (already existing) `translateResponseCartToWPCOMCart` function, which turns the `ReponseCart` into a `WPCOMCart`.

As a side effect, this PR also splits up the different parts of the hook into separate files to make them easier to read and modify.

#### API changes

Before this PR, the usage of the hook looks like this:

```js
const {
	items,
	tax,
	couponItem,
	total,
	credits,
	removeItem,
	submitCoupon,
	removeCoupon,
	updateLocation,
	couponStatus,
	changeItemVariant,
	errors,
	subtotal,
	isLoading,
	isPendingUpdate,
	allowedPaymentMethods,
	variantSelectOverride,
	responseCart,
	addItem,
} = useShoppingCart(
	siteSlug,
	canInitializeCart,
	productsForCart,
	couponCodeFromUrl,
	setCart,
	getCart,
	showAddCouponSuccessMessage,
	recordEvent
);
```

After this PR, the usage of the hook looks like this:

```js
const {
	removeItem,
	couponStatus,
	submitCoupon,
	removeCoupon,
	updateLocation,
	changeItemVariant,
	isLoading,
	isPendingUpdate,
	responseCart,
	addItem,
	variantSelectOverride,
} = useShoppingCartManager( {
	cartKey,
	canInitializeCart,
	productsToAdd,
	couponToAdd,
	setCart,
	getCart,
	onEvent,
} );
```

#### Testing instructions

This is a fairly large modification and touches many aspects of using checkout, but the good news is that it is so central to checkout that any breakages should be pretty obvious. Some notable things to test to make sure they work as expected:
 
- Plan variation changes change plan as expected.
- Adding and removing coupons works as expected and you see the "The coupon was added" success notification.
- Adding a product via the URL works (if one product works, all of them will, so no need to test the variations).
- The form shows a loading state while loading.
- The form is disabled while the cart updates.
- Line items are displayed.
- Tax is displayed (try US postal code 10001).
- Total is displayed.
- Credits are displayed (you'll need to add some).
- Subtotal is displayed.
- Different payment methods are displayed (existing cards, credit card, and paypal should suffice).
